### PR TITLE
Event API

### PR DESCRIPTION
**Pull request by mfranciszkiewicz**

### CODE REVIEW

**Review by Wiezzel — DISMISSED** [review]

**Review by nieznanysprawiciel — APPROVED** [review]
I assumed, that all changes in `event-api/master` were reviewed, before merging to this branch.

**Comment by tworec** [comment]
> I assumed, that all changes in `event-api/master` were reviewed, before merging to this branch.

not true ;)

File: Cargo.toml [hunk 1]
```diff
@@ -171,8 +172,11 @@ ya-sb-router = { path = "service-bus/router" }
 ya-sb-util = { path = "service-bus/util" }
 
 ## CLIENT
-ya-client = { git = "https://github.com/golemfactory/ya-client.git", rev = "e68bbd90c903c6f74f77fddb80a432d20e4e2b2a"}
-ya-client-model = { git = "https://github.com/golemfactory/ya-client.git", rev = "e68bbd90c903c6f74f77fddb80a432d20e4e2b2a"}
+ya-client = { git = "https://github.com/golemfactory/ya-client.git", branch = "event-api/master"}
```
Comment (tworec) [inline Cargo.toml:175]: we do not want to merge it like this to the master branch!

**Review by tworec — CHANGES_REQUESTED** [review]
need to point specific ya-client gitrev

File: core/activity/src/common.rs [hunk 2]
```diff
@@ -39,10 +39,15 @@ pub struct QueryTimeoutCommandIndex {
 }
 
 #[derive(Deserialize, Debug)]
-pub struct QueryTimeoutMaxEvents {
+pub struct QueryEvents {
     /// number of milliseconds to wait
-    #[serde(rename = "timeout", default = "default_query_timeout")]
-    pub timeout: Option<f32>,
+    #[serde(rename = "appSessionId")]
```
Comment (tworec) [inline core/activity/src/common.rs:44]: above comment is irrelevant
cc @mfranciszkiewicz @nieznanysprawiciel

Comment (tworec) [reply core/activity/src/common.rs:44]: forthermore I'd use `#[serde(rename_all = "camelCase")]` which is not only more concise, but also less error-prone

Comment (tworec) [reply core/activity/src/common.rs:44]: might be adressed in separate PR

Comment (mfranciszkiewicz) [reply core/activity/src/common.rs:44]: afair this code was auto-generated and only slightly modified later on

File: core/activity/src/dao/event.rs [hunk 3]
```diff
@@ -99,65 +106,66 @@ impl<'c> EventDao<'c> {
...
                 .order(dsl_event::event_date.asc())
-                .limit(limit as i64)
+                .limit(limit)
                 .load::<Event>(conn)
                 .optional()?;
 
-            if let Some(r) = &results {
-                let ids = r.iter().map(|event| event.id).collect::<Vec<_>>();
-                if !ids.is_empty() {
-                    diesel::delete(dsl_event::activity_event.filter(dsl_event::id.eq_any(ids)))
```
Comment (tworec) [inline core/activity/src/dao/event.rs:135]: we need some cleanup mechanism for those events

**Review by Wiezzel — DISMISSED** [review]

File: core/activity/src/dao/event.rs [hunk 4]
```diff
@@ -99,65 +106,66 @@ impl<'c> EventDao<'c> {
...
 
         log::trace!("get_events: starting db query");
-        let identity_id = identity_id.to_owned();
-
-        do_with_transaction(self.pool, move |conn| {
-            let results: Option<Vec<Event>> = dsl_event::activity_event
+        readonly_transaction(self.pool, move |conn| {
+            let mut query = dsl_event::activity_event
                 .inner_join(schema::activity::table)
-                .filter(dsl_event::identity_id.eq(identity_id))
```
Comment (tworec) [inline core/activity/src/dao/event.rs:119]: is it on purpose to skip identity filtering here? it seems that now every identity can sniff all events, even those not intended for it

File: core/market/src/db/dao/negotiation_events.rs [hunk 5]
```diff
@@ -40,17 +40,28 @@ impl<'c> AsDao<'c> for NegotiationEventsDao<'c> {
 }
 
 impl<'c> NegotiationEventsDao<'c> {
-    pub async fn add_proposal_event(
+    pub async fn add_proposal_event(&self, proposal: &Proposal, role: Owner) -> DbResult<()> {
+        let event = MarketEvent::from_proposal(proposal, role);
```
Comment (tworec) [inline core/market/src/db/dao/negotiation_events.rs:44]: we need to create it under the transaction and with a timestamp on the rust code side, sam as we did with agreement events

File: core/market/src/db/dao/negotiation_events.rs [hunk 6]
```diff
@@ -40,17 +40,28 @@ impl<'c> AsDao<'c> for NegotiationEventsDao<'c> {
...
+
+    pub async fn add_proposal_rejected_event(
         &self,
-        proposal: Proposal,
-        owner: OwnerType,
-    ) -> DbResult<Proposal> {
+        proposal: &Proposal,
+        reason: Option<String>,
+    ) -> DbResult<()> {
+        let event = MarketEvent::proposal_rejected(proposal, reason);
```
Comment (tworec) [inline core/market/src/db/dao/negotiation_events.rs:59]: same here

File: core/activity/src/dao/event.rs (hunk 4 above)
Comment (mfranciszkiewicz) [reply core/activity/src/dao/event.rs:119]: reverted

File: core/activity/src/dao/event.rs (hunk 3 above)
Comment (mfranciszkiewicz) [reply core/activity/src/dao/event.rs:135]: Do you mean events in general? The count of activity events is going to be few orders of magnitude lower than payments / market

File: core/market/src/negotiation/common.rs [hunk 7]
```diff
@@ -125,12 +133,56 @@ impl CommonBroker {
...
+        let proposal = self.get_proposal(subs_id.clone(), proposal_id).await?;
+
+        self.validate_proposal(&proposal, caller_id, caller_role)
+            .await?;
+
+        if proposal.body.issuer == Issuer::Us && proposal_id.owner() == caller_role {
+            let e = ProposalValidationError::OwnProposal(proposal.body.id.clone());
+            log::warn!("{}", e);
+            Err(e)?;
+        }
```
Comment (tworec) [inline core/market/src/negotiation/common.rs:159]: I've written it myself, but it smells to me. This should be part of `validate_proposal`

Comment (tworec) [reply core/market/src/negotiation/common.rs:159]: addressed in #914

File: core/market/src/negotiation/common.rs [hunk 8]
```diff
@@ -316,53 +366,48 @@ impl CommonBroker {
...
             .await
             .map_err(|e| AgreementError::UpdateState((&agreement.id).clone(), e))?;
-
         self.notify_agreement(&agreement).await;
 
         inc_terminate_metrics(&reason, agreement.id.owner());
         log::info!(
-            "Agent {} terminated Agreement [{}]. Reason: {}",
+            "{:?} {} terminated Agreement [{}]. Reason: {}",
+            agreement.id.owner(),
```
Comment (tworec) [inline core/market/src/negotiation/common.rs:379]: minor: we are not sure here that the owner is terminating

File: core/payment/src/api.rs [hunk 9]
```diff
@@ -5,22 +5,27 @@ use ya_client_model::payment::PAYMENT_API_PATH;
...
 }
 
 pub fn web_scope(db: &DbExecutor) -> Scope {
     Scope::new(PAYMENT_API_PATH)
         .data(db.clone())
-        .service(provider_scope())
-        .service(requestor_scope())
+        .service(api_scope(Scope::new("")))
+    // TODO: TEST
+    // Scope::new(PAYMENT_API_PATH).extend(api_scope).data(db.clone())
```
Comment (tworec) [inline core/payment/src/api.rs:28]: both defaults below should be 5 seconds according to actual spec:
- pollTmeout: https://github.com/golemfactory/ya-client/blob/87f106498a7a7b1519d82d0981fc21b260a89cf2/specs/common.yaml#L18
- ackTimeout: https://github.com/golemfactory/ya-client/blob/87f106498a7a7b1519d82d0981fc21b260a89cf2/specs/common.yaml#L32

File: core/activity/src/dao/event.rs [hunk 10]
```diff
@@ -64,14 +64,14 @@ impl<'c> EventDao<'c> {
...
 
-        let now = Utc::now().naive_utc();
         log::trace!("creating event_type: {:?}", event_type);
 
         let app_session_id = app_session_id.to_owned();
         let activity_id = activity_id.to_owned();
         let identity_id = identity_id.to_owned();
 
         do_with_transaction(self.pool, move |conn| {
+            let now = Utc::now().naive_utc();
```
Comment (tworec) [inline core/activity/src/dao/event.rs:74]: 👍

**Review by tworec — APPROVED** [review]
minor issues only, all of which can and should be addressed in separate PRs

File: core/activity/src/dao/event.rs (hunk 3 above)
Comment (tworec) [reply core/activity/src/dao/event.rs:135]: I think like, at most one order of magnitude.
But yes, in general.
We have a cleanup mechanism in mkt

Comment (mfranciszkiewicz) [reply core/activity/src/dao/event.rs:135]: | I think like, at most one order of magnitude.

E.g. debit note events may significantly outnumber create / destroy activity events for long running services.

| we need some cleanup mechanism for those events

We haven't established a common cleanup policy and event removal isn't a part of the spec (for whatever reason). I'll abstain from implementing a heuristic solution for now - there is simply no need to do that now and the number of events won't affect the database in a significant way

Comment (tworec) [reply core/activity/src/dao/event.rs:135]: True dat!
I was not clear in my intention here. I just wanted to express that such cleanup is needed, but not pushing to implement it right away. 👍

Comment (tworec) [reply core/activity/src/dao/event.rs:135]: referenced it #918

File: core/market/src/db/dao/negotiation_events.rs (hunk 5 above)
Comment (tworec) [reply core/market/src/db/dao/negotiation_events.rs:44]: referenced in #917

File: core/payment/src/api.rs (hunk 9 above)
Comment (tworec) [reply core/payment/src/api.rs:28]: referenced in #919

**Pull request merged**